File: cpp/src/transport.cpp
// RFCOMM transport implementation using raw Linux Bluetooth sockets.
#include "transport.h"

#ifndef __APPLE__
#include <bluetooth/bluetooth.h>
#include <bluetooth/rfcomm.h>
#include <cerrno>
#include <cstring>
#include <sys/socket.h>
#include <thread>
#include <unistd.h>

namespace bmap {

static void parse_mac(const std::string& mac, bdaddr_t& addr) {
    if (str2ba(mac.c_str(), &addr) < 0) {
        throw std::runtime_error("Invalid MAC address: " + mac);
    }
}

static void set_timeout(int fd, int opt, int ms) {
    struct timeval tv;
    tv.tv_sec = ms / 1000;
    tv.tv_usec = (ms % 1000) * 1000;
    if (setsockopt(fd, SOL_SOCKET, opt, &tv, sizeof(tv)) < 0) {
        throw std::runtime_error(std::string("setsockopt failed: ") + strerror(errno));
    }
}

RfcommTransport::RfcommTransport(const std::string& mac, uint8_t channel) {
    fd_ = socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM);
    if (fd_ < 0) {
        throw std::runtime_error(std::string("Failed to create socket: ") + strerror(errno));
    }

    set_timeout(fd_, SO_SNDTIMEO, 3000);

    struct sockaddr_rc addr{};
    addr.rc_family = AF_BLUETOOTH;
    parse_mac(mac, addr.rc_bdaddr);
    addr.rc_channel = channel;

    if (connect(fd_, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr)) < 0) {
        ::close(fd_);
        fd_ = -1;
        throw std::runtime_error("Failed to connect to " + mac + ": " + strerror(errno));
    }

    set_timeout(fd_, SO_RCVTIMEO, 3000);
}

RfcommTransport::~RfcommTransport() {
    if (fd_ >= 0) ::close(fd_);
}

std::vector<uint8_t> RfcommTransport::send_recv(const std::vector<uint8_t>& packet) {
    return send_recv_inner(packet, false);
}

std::vector<uint8_t> RfcommTransport::send_recv_drain(const std::vector<uint8_t>& packet) {
    return send_recv_inner(packet, true);
}

std::vector<uint8_t> RfcommTransport::send_recv_inner(const std::vector<uint8_t>& packet, bool drain) {
    // Send
    ssize_t sent = ::send(fd_, packet.data(), packet.size(), 0);
    if (sent < 0) {
        throw std::runtime_error(std::string("Send failed: ") + strerror(errno));
    }

    // Protocol-required delay for device processing.
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    // Receive
    uint8_t buf[4096];
    ssize_t n = ::recv(fd_, buf, sizeof(buf), 0);
    if (n <= 0) {
        throw std::runtime_error(std::string("No response: ") + strerror(errno));
    }
    std::vector<uint8_t> data(buf, buf + n);

    if (drain) {
        set_recv_timeout(500);
        while (true) {
            n = ::recv(fd_, buf, sizeof(buf), 0);
            if (n <= 0) break;
            data.insert(data.end(), buf, buf + n);
        }
        set_recv_timeout(3000);
    }

    return data;
}

void RfcommTransport::set_recv_timeout(int ms) {
    // Best-effort for drain loop.
    struct timeval tv;
    tv.tv_sec = ms / 1000;
    tv.tv_usec = (ms % 1000) * 1000;
    setsockopt(fd_, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
}

} // namespace bmap

#else // __APPLE__
#include <stdexcept>

namespace bmap {

RfcommTransport::RfcommTransport(const std::string&, uint8_t) {
    throw std::runtime_error("Bluetooth RFCOMM sockets not supported on macOS in C++ library. Use Python bosectl instead.");
}

RfcommTransport::~RfcommTransport() {}

std::vector<uint8_t> RfcommTransport::send_recv(const std::vector<uint8_t>&) {
    throw std::runtime_error("Not supported on macOS");
}

std::vector<uint8_t> RfcommTransport::send_recv_drain(const std::vector<uint8_t>&) {
    throw std::runtime_error("Not supported on macOS");
}

std::vector<uint8_t> RfcommTransport::send_recv_inner(const std::vector<uint8_t>&, bool) {
    throw std::runtime_error("Not supported on macOS");
}

void RfcommTransport::set_recv_timeout(int) {}

} // namespace bmap
#endif

File: rust/src/transport.rs
//! RFCOMM Bluetooth socket transport for BMAP devices.
//!
//! Uses raw Linux AF_BLUETOOTH sockets via libc — no external
//! Bluetooth library needed. Mirrors what Python's socket.AF_BLUETOOTH
//! does under the hood.

use std::io;
use std::os::fd::{FromRawFd, OwnedFd, AsRawFd};
use std::time::Duration;

use crate::error::{BmapError, BmapResult};

const AF_BLUETOOTH: i32 = 31;
const BTPROTO_RFCOMM: i32 = 3;

/// RFCOMM channel for BMAP protocol.
pub const BMAP_CHANNEL: u8 = 2;

/// Kernel sockaddr_rc structure (matches `struct sockaddr_rc` in bluetooth.h).
///
/// Note: The kernel struct is `__attribute__((packed))` at 9 bytes, but
/// `repr(C)` on x86-64 produces 10 bytes (1 byte padding after rc_channel
/// for u16 alignment). We pass `size_of::<SockaddrRc>()` to connect(),
/// and the kernel accepts both 9 and 10 for BTPROTO_RFCOMM.
#[repr(C)]
struct SockaddrRc {
    rc_family: u16,
    rc_bdaddr: [u8; 6],
    rc_channel: u8,
}

/// Parse a "AA:BB:CC:DD:EE:FF" MAC string into bytes (reversed for BlueZ).
fn parse_mac(mac: &str) -> BmapResult<[u8; 6]> {
    let parts: Vec<&str> = mac.split(':').collect();
    if parts.len() != 6 {
        return Err(BmapError::Connection(format!("Invalid MAC address: {}", mac)));
    }
    let mut addr = [0u8; 6];
    for (i, part) in parts.iter().enumerate() {
        addr[5 - i] = u8::from_str_radix(part, 16)
            .map_err(|_| BmapError::Connection(format!("Invalid MAC byte: {}", part)))?;
    }
    Ok(addr)
}

/// Transport abstraction for sending/receiving BMAP packets.
///
/// Implemented by `RfcommTransport` for real Bluetooth. Can be mocked for testing.
pub trait Transport {
    fn send_recv(&self, packet: &[u8]) -> BmapResult<Vec<u8>>;
    fn send_recv_drain(&self, packet: &[u8]) -> BmapResult<Vec<u8>>;
}

/// Raw RFCOMM Bluetooth socket transport.
pub struct RfcommTransport {
    fd: OwnedFd,
}

impl RfcommTransport {
    /// Connect to a BMAP device by MAC address and channel.
    pub fn connect(mac: &str, channel: u8) -> BmapResult<Self> {
        let bdaddr = parse_mac(mac)?;

        unsafe {
            let fd = libc::socket(AF_BLUETOOTH, libc::SOCK_STREAM, BTPROTO_RFCOMM);
            if fd < 0 {
                return Err(BmapError::Connection(format!(
                    "Failed to create socket: {}", io::Error::last_os_error()
                )));
            }
            let owned = OwnedFd::from_raw_fd(fd);

            let addr = SockaddrRc {
                rc_family: AF_BLUETOOTH as u16,
                rc_bdaddr: bdaddr,
                rc_channel: channel,
            };

            // Set connect timeout
            set_sock_timeout(owned.as_raw_fd(), libc::SO_SNDTIMEO, Duration::from_secs(3))?;

            let ret = libc::connect(
                owned.as_raw_fd(),
                &addr as *const _ as *const libc::sockaddr,
                std::mem::size_of::<SockaddrRc>() as u32,
            );
            if ret < 0 {
                return Err(BmapError::Connection(format!(
                    "Failed to connect to {}: {}", mac, io::Error::last_os_error()
                )));
            }

            // Set recv timeout
            set_sock_timeout(owned.as_raw_fd(), libc::SO_RCVTIMEO, Duration::from_secs(3))?;

            Ok(Self { fd: owned })
        }
    }

    fn send_recv_inner(&self, packet: &[u8], drain: bool) -> BmapResult<Vec<u8>> {
        let fd = self.fd.as_raw_fd();

        // Send — use libc::send directly to avoid fd ownership issues.
        let sent = unsafe {
            libc::send(fd, packet.as_ptr() as *const libc::c_void, packet.len(), 0)
        };
        if sent < 0 {
            return Err(BmapError::Connection(format!(
                "Send failed: {}", io::Error::last_os_error()
            )));
        }

        // Brief delay for device to process (protocol-required).
        std::thread::sleep(Duration::from_millis(200));

        // Receive
        let mut buf = [0u8; 4096];
        let n = unsafe {
            libc::recv(fd, buf.as_mut_ptr() as *mut libc::c_void, buf.len(), 0)
        };
        if n <= 0 {
            return Err(BmapError::Timeout(format!(
                "No response: {}", io::Error::last_os_error()
            )));
        }
        let mut data = buf[..n as usize].to_vec();

        if drain {
            self.set_recv_timeout(Duration::from_millis(500));
            loop {
                let n = unsafe {
                    libc::recv(fd, buf.as_mut_ptr() as *mut libc::c_void, buf.len(), 0)
                };
                if n <= 0 {
                    break;
                }
                data.extend_from_slice(&buf[..n as usize]);
            }
            self.set_recv_timeout(Duration::from_secs(3));
        }

        Ok(data)
    }

    fn set_recv_timeout(&self, duration: Duration) {
        // Best-effort — drain loop still works if this fails.
        let _ = unsafe { set_sock_timeout(self.fd.as_raw_fd(), libc::SO_RCVTIMEO, duration) };
    }
}

/// Set a socket timeout, checking the return value.
unsafe fn set_sock_timeout(fd: i32, opt: i32, duration: Duration) -> BmapResult<()> {
    let timeout = libc::timeval {
        tv_sec: duration.as_secs() as _,
        tv_usec: duration.subsec_micros() as _,
    };
    let ret = libc::setsockopt(
        fd,
        libc::SOL_SOCKET,
        opt,
        &timeout as *const _ as *const libc::c_void,
        std::mem::size_of::<libc::timeval>() as u32,
    );
    if ret < 0 {
        return Err(BmapError::Connection(format!(
            "setsockopt failed: {}", io::Error::last_os_error()
        )));
    }
    Ok(())
}

impl Transport for RfcommTransport {
    fn send_recv(&self, packet: &[u8]) -> BmapResult<Vec<u8>> {
        self.send_recv_inner(packet, false)
    }

    fn send_recv_drain(&self, packet: &[u8]) -> BmapResult<Vec<u8>> {
        self.send_recv_inner(packet, true)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_mac() {
        let addr = parse_mac("AA:BB:CC:DD:EE:FF").unwrap();
        // BlueZ stores MAC in reverse byte order
        assert_eq!(addr, [0xFF, 0xEE, 0xDD, 0xCC, 0xBB, 0xAA]);
    }

    #[test]
    fn test_parse_mac_invalid() {
        assert!(parse_mac("not-a-mac").is_err());
        assert!(parse_mac("AA:BB:CC").is_err());
        assert!(parse_mac("GG:HH:II:JJ:KK:LL").is_err());
    }

    #[cfg(target_os = "macos")]
    #[test]
    fn test_macos_transport_fails() {
        let res = RfcommTransport::connect("00:11:22:33:44:55", 2);
        assert!(res.is_err());
    }
}
